### DISCUSSION TITLE
Config Management Enhancements

### DISCUSSION THREAD

**Discussion by kralicky** [rancher/opni#1008] [Ideas]
We need to design a system to consolidate all the gateway, agent, and plugin config files (and maybe configuration for other opni components) in a single place. I want to start a discussion and gather requirements from different teams so we can plan a design for an enhanced config system. 

Features needed for monitoring:
- Hot-reload with callbacks
- A clean way to get config objects into plugins, preferably without fetching them directly through the management api
- An API for updating config fields for the UI/CLI to use

Unknowns:
- Where to store config files?
- How to structure configs?
- How should hot-reloading work?
- Should we define the config with protobuf?
- Are there any libraries we can use?
- How does the KV store fit into this?

**Comment by dbason** [comment]
It might be worth checking into how Rancher stores its configuration currently.  I would tend to storing it in a k8s object, but not sure if this would best be a json blob stored in a ConfigMap or a custom resource.

**Reply by sanjay920** [reply]
https://ranchermanager.docs.rancher.com/getting-started/overview#features-of-the-rancher-api-server

**Reply by dbason** [reply]
Yeah I meant more from a code point of view.  More specifically how it stores these settings:
![image](https://user-images.githubusercontent.com/8053452/215909743-9f64b35c-4047-43e3-a5a4-8ca01d358aa4.png)

**Comment by alexandreLamarre** [comment]
I'd be in favour of looking for a library that programmatically creates config file(s) with developer friendly methods, independent of the physical formats its translated to. I was using something similar in rust a while back.

Not sure if such a go library exists, we tend to be very reliant in go on json, yaml , kubebuilder, etc... tags for something like that

**Reply by kralicky** [reply]
Which Rust library is that?
